File: packages/csharp/RealtimeMail/RealtimeMailClient.cs
namespace RealtimeMail;

public sealed class RealtimeMailClient
{
    public RealtimeMailClient(
        ManifestResolver manifests,
        TrustPolicy trust,
        RealtimeGatewayClient gateway,
        TraditionalMailAccountManager accounts
    )
    {
        Manifests = manifests;
        Trust = trust;
        Gateway = gateway;
        Accounts = accounts;
    }

    public ManifestResolver Manifests { get; }

    public TrustPolicy Trust { get; }

    public RealtimeGatewayClient Gateway { get; }

    public TraditionalMailAccountManager Accounts { get; }

    public Task<string> DiscoverAsync(string domain, CancellationToken cancellationToken = default)
    {
        return Manifests.ResolveJsonAsync(domain, cancellationToken);
    }

    public void TrustDomain(string domain)
    {
        Trust.TrustDomain(domain);
    }

    public Task<ISubscription> SubscribeAsync(
        RealtimeMailChannel channel,
        Action<RealtimeMailMessage> onMessage,
        CancellationToken cancellationToken = default
    )
    {
        return Gateway.SubscribeAsync(channel, onMessage, cancellationToken);
    }
}

File: packages/java/src/main/java/org/realtimemail/RealtimeMailClient.java
package org.realtimemail;

import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

public final class RealtimeMailClient {
  private final ManifestResolver manifests;
  private final TrustPolicy trust;
  private final RealtimeGatewayClient gateway;
  private final TraditionalMailAccountManager accounts;

  public RealtimeMailClient(
    ManifestResolver manifests,
    TrustPolicy trust,
    RealtimeGatewayClient gateway,
    TraditionalMailAccountManager accounts
  ) {
    this.manifests = manifests;
    this.trust = trust;
    this.gateway = gateway;
    this.accounts = accounts;
  }

  public CompletableFuture<String> discover(String domain) {
    return manifests.resolveJson(domain);
  }

  public void trustDomain(String domain) {
    trust.trustDomain(domain);
  }

  public CompletableFuture<Subscription> subscribe(
    RealtimeMailChannel channel,
    Consumer<RealtimeMailMessage> onMessage
  ) {
    return gateway.subscribe(channel, onMessage);
  }

  public TraditionalMailAccountManager accounts() {
    return accounts;
  }
}
